sdks/go/pkg/beam/core/util/hooks/hooks.go

```diff
@@ -0,0 +1,184 @@
+type HookFactory func([]string) Hook
+
+// RegisterHook registers a Hook for the
+// supplied identifier.
+func RegisterHook(name string, h HookFactory) {
+	hookRegistry[name] = h
+}
+
+// RunInitHooks runs the init hooks.
+func RunInitHooks(ctx context.Context) error {
+	// If an init hook fails to complete, the invariants of the
+	// system are compromised and we can't run a workflow.
+	// The hooks can run in any order. They should not be
+	// interdependent or interfere with each other.
+	for _, h := range activeHooks {
+		if h.Init != nil {
+			if err := h.Init(ctx); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// RequestHook is called when handling a Fn API instruction.
+type RequestHook func(context.Context, *fnpb.InstructionRequest) error
+
+// RunRequestHooks runs the hooks that handle a FnAPI request.
+func RunRequestHooks(ctx context.Context, req *fnpb.InstructionRequest) {
+	// The request hooks should not modify the request.
+	// TODO(wcn): pass the request by value to enforce? That's a perf hit.
+	// I'd rather trust users to do the right thing.
+	for n, h := range activeHooks {
+		if h.Req != nil {
+			if err := h.Req(ctx, req); err != nil {
+				log.Infof(ctx, "request hook %s failed: %v", n, err)
+			}
+		}
+	}
+}
+
+// ResponseHook is called when sending a Fn API instruction response.
+type ResponseHook func(context.Context, *fnpb.InstructionRequest, *fnpb.InstructionResponse) error
+
+// RunResponseHooks runs the hooks that handle a FnAPI response.
+func RunResponseHooks(ctx context.Context, req *fnpb.InstructionRequest, resp *fnpb.InstructionResponse) {
+	for n, h := range activeHooks {
+		if h.Resp != nil {
+			if err := h.Resp(ctx, req, resp); err != nil {
+				log.Infof(ctx, "response hook %s failed: %v", n, err)
+			}
+		}
+	}
+}
+
+// SerializeHooks serializes the activated hooks and their configuration into a JSON string
+// that can be deserialized later by the runner.
+func SerializeHooks() {
+	data, err := json.Marshal(enabledHooks)
+	if err != nil {
+		// Shouldn't happen, since all the data is strings.
+		panic(fmt.Sprintf("Couldn't serialize hooks: %v", err))
+	}
+	fmt.Fprintf(os.Stderr, "SerializeHooks: %q", string(data))
+	runtime.GlobalOptions.Set("hooks", string(data))
+}
+
+// DeserializeHooks extracts the hook configuration information from the options and configures
+// the hooks with the supplied options.
+func DeserializeHooks() {
+	cfg := runtime.GlobalOptions.Get("hooks")
+	if err := json.Unmarshal([]byte(cfg), &enabledHooks); err != nil {
+		// Shouldn't happen
+		panic(fmt.Sprintf("DeserializeHooks failed on input %q: %v", cfg, err))
+	}
+
+	for h, opts := range enabledHooks {
+		activeHooks[h] = hookRegistry[h](opts)
+	}
+}
+
+// EnableHook enables the hook to be run for the pipline. It will be
+// receive the supplied args when the pipeline executes. It is safe
+// to enable the same hook with different options, as this is necessary
+// if a hook wants to compose behavior.
```

Comment:
<!--new_thread; commit:5d13d3df224a378bb27856f0351e23e0898e862b; resolved:0-->
When I'm reading "compose behavior", it sounds to me that the same hook will run with twice with the same args, but as implemented it's the last EnableHook call for the name that wins.

If that's the intended behaviour for the call, please include it in the comment.

Reply:
Reworded. It's the last call to EnableHook that wins.